### DISCUSSION TITLE
Upcasting events while replaying

### DISCUSSION THREAD

**Discussion by rogierknoester** [serverlesstechnology/cqrs#79] [Q&A] [answered]
Let me preface, I am relatively new to ES.

I have a project where after a few weeks I've decided I want to change some events' structure. The new data can be added by using the upcasting feature of this framework. Of course, the existing read models would also need this data. Then I would replay existing events. But I am not running into the issue that it seems like the `QueryReplay::replay_all()` and inside there the `PersistedEventRepository::stream_all_events` & `ReplayStream` do currently not use upcasters. Nor is there a way to provide them. Other `PersistedEventRepository` functions do consider the (optionally) passed upcasters and apply them.

Am I wrong in my perception that during replay the upcasting should also happen? It feels like, right now, the best way would be to load events from the `PersistedEventStore` (instead of repository) per aggregate and upcast its events and manually replay them onto queries/projections.

**Comment by davegarred** [comment, marked as answer]
Hi @rogierknoester, you're correct that when replaying events you'll want to apply those upcasters. I think that would be a nice feature, I've added this as [an issue](https://github.com/serverlesstechnology/cqrs/issues/80). Thanks for the feedback!

**Reply by rogierknoester** [reply]
Thanks for the swift reply and good to hear it should be there! I can probably reserve some time and submit a PR for it, would that be useful?

**Reply by davegarred** [reply]
Absolutely, take a shot at it.
